### ISSUE TITLE
What dose it mean by REPR FAILED?

### ISSUE THREAD

**Issue by nuomizai** [zasdfgbnm/TorchSnooper#20]
Thanks for the contributer, I got this excellent debug tool for my pytorch code. But I've got a problem when using torchsnooper.snoop(). The terminal information is like:

15:38:36.750755 line       120             action_cat_batch = self.action_cat[indices]
New var:....... action_cat_batch = REPR FAILED

But my action_cat_batch is nothing different from the other var, for example the obs_batch can be shown correctly:

ew var:....... obs_batch = tensor<(125, 3, 80, 80), float32, cuda:0>
15:38:36.743800 line       119             next_obs_batch = self.obs[next_indices]

And my debug code is like this:

    @torchsnooper.snoop()
    def feed_forward_generator(self, advantages, rank):
        mini_batch_size = self.num_steps // self.mini_batch_num
        sampler = BatchSampler(SubsetRandomSampler(range(self.num_steps)), mini_batch_size, drop_last=False)
        for indices in sampler:
            next_indices = [indice + 1 for indice in indices]
            # [index, dim]
            obs_batch = self.obs[indices]
            next_obs_batch = self.obs[next_indices]
            action_cat_batch = self.action_cat[indices]
            action_dia_batch = self.action_dia[indices].transpose(0,1)
            value_pred_batch = self.value_preds[indices]
            return_batch = self.returns[indices]
            old_action_log_probs_batch = self.action_log_probs[indices]
            advantages_batch = advantages[indices]
            masks_batch = self.masks[indices]

            yield obs_batch, next_obs_batch, (action_cat_batch, action_dia_batch), value_pred_batch, return_batch, \
                  masks_batch, old_action_log_probs_batch, advantages_batch

I don't quite understand what dose it mean by 'REPR FAILED', can anyone help me?

**Comment by nuomizai** [comment]
And the initial code is like this:
    def __init__(self, num_steps, mini_batch_num, obs_shape, uav_num):
        self.mini_batch_num = mini_batch_num
        self.obs = torch.zeros(num_steps + 1, *obs_shape)
        self.next_obs = torch.zeros(num_steps + 1, *obs_shape)
        self.rewards = torch.zeros(num_steps, 1)
        self.value_preds = torch.zeros(num_steps + 1, 1)
        self.returns = torch.zeros(num_steps + 1, 1)
        self.action_log_probs = torch.zeros(num_steps, 1)
        self.action_cat = torch.zeros(num_steps, 1, dtype=torch.long)
        self.action_dia = torch.zeros(num_steps, int(uav_num * 2), dtype=torch.long)
        self.masks = torch.ones(num_steps + 1, 1)
        self.num_steps = num_steps
        self.step = 0

I'm guessing that if the tensor's data type makes the difference. How can I repair that?

**Comment by alexmojaki** [comment]
Try this:

```python
import torchsnooper
import snoop
from cheap_repr import cheap_repr

torchsnooper.register_snoop()
cheap_repr.raise_exceptions = True

@snoop
def feed_forward_generator(self, advantages, rank):
    ...
```

**Comment by zasdfgbnm** [comment]
@nuomizai `REPR FAILED` means when trying to convert your object (probably a tensor) into its string representation, an error happens.

**Comment by nuomizai** [comment]
Thank you, Alex. I try the snoop method and a Runtime Error happened:

RuntimeError: value cannot be converted to type int16_t without overflow: inf

I don't quiet understand how this happens. Because without the snoop method, this code can run correctly.

**Comment by nuomizai** [comment]
@alexmojaki

**Comment by nuomizai** [comment]
@zasdfgbnm Thanks for your reply! Do you know how to fix this error? Because without the @snoop method, this code can run without bug reporting.

**Comment by zasdfgbnm** [comment]
Hi @nuomizai, could please post the whole error traceback? That would be more helpful to find what's wrong.

**Comment by nuomizai** [comment]
Sure, the error traceback is like this:

22:26:38.66  123 |             next_indices = [indice + 1 for indice in indices]
22:26:38.66 .................. next_indices = [29, 449, 274, 441, 457, 412, 424, 433, 221, 51, ..., 286, 361, 447, 203, 330, 46, 310, 419, 105, 442]
22:26:38.66  125 |             obs_batch = self.obs[indices]
22:26:38.66 .................. obs_batch = tensor<(125, 3, 80, 80), float32, cuda:0>
22:26:38.66  126 |             next_obs_batch = self.obs[next_indices]
22:26:38.66 .................. next_obs_batch = tensor<(125, 3, 80, 80), float32, cuda:0>
22:26:38.66  127 |             action_cat_batch = self.action_cat[indices]
Process Process-5:
Traceback (most recent call last):
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/linc/zhaoyinuo/experiment_2/uav2_charge2/exper_dppo/train.py", line 170, in train
    for samples in data_generator:
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/snoop/tracer.py", line 178, in generator_wrapper
    outgoing = method(incoming)
  File "/home/linc/zhaoyinuo/experiment_2/uav2_charge2/exper_dppo/storage.py", line 128, in feed_forward_generator
    action_dia_batch = self.action_dia[indices].transpose(0, 1)
  File "/home/linc/zhaoyinuo/experiment_2/uav2_charge2/exper_dppo/storage.py", line 128, in feed_forward_generator
    action_dia_batch = self.action_dia[indices].transpose(0, 1)
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/snoop/tracer.py", line 266, in trace
    self.variable_whitelist,
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/snoop/tracer.py", line 53, in update_variables
    self.get_local_reprs(watch, watch_extras, whitelist)
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/snoop/tracer.py", line 52, in <genexpr>
    for source, value in
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/snoop/utils.py", line 90, in my_cheap_repr
    return cheap_repr(x, target_length=REPR_TARGET_LENGTH)
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/cheap_repr/__init__.py", line 159, in cheap_repr
    return _try_repr(func, x, helper)
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/cheap_repr/__init__.py", line 172, in _try_repr
    return func(x, *args)
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/torchsnooper/__init__.py", line 160, in <lambda>
    cheap_repr.register_repr(torch.Tensor)(lambda x, _: tensor_format(x))
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/torchsnooper/__init__.py", line 58, in __call__
    result = bool(torch.isinf(tensor).any())
  File "/home/linc/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/functional.py", line 287, in isinf
    return tensor.abs() == inf
RuntimeError: value cannot be converted to type int16_t without overflow: inf****

**Comment by nuomizai** [comment]
@zasdfgbnm  And this is the whole class, hopefully this can help.
import torch
from torch.utils.data.sampler import BatchSampler, SubsetRandomSampler
from util.self_define_utils import self_flatten
import torchsnooper
import snoop
from cheap_repr import cheap_repr

torchsnooper.register_snoop()
cheap_repr.raise_exceptions = True


class RolloutStorage(object):
    def __init__(self, num_steps, mini_batch_num, obs_shape, uav_num):
        self.mini_batch_num = mini_batch_num
        self.obs = torch.zeros(num_steps + 1, *obs_shape)
        self.next_obs = torch.zeros(num_steps + 1, *obs_shape)
        self.rewards = torch.zeros(num_steps, 1)
        self.value_preds = torch.zeros(num_steps + 1, 1)
        self.returns = torch.zeros(num_steps + 1, 1)
        self.action_log_probs = torch.zeros(num_steps, 1)
        self.action_cat = torch.zeros(num_steps, 1, dtype=torch.int16)
        self.action_dia = torch.zeros(num_steps, int(uav_num * 2), dtype=torch.int16)
        self.masks = torch.ones(num_steps + 1, 1)
        self.num_steps = num_steps
        self.step = 0

    def to(self, device):
        self.obs = self.obs.to(device)
        self.rewards = self.rewards.to(device)
        self.value_preds = self.value_preds.to(device)
        self.returns = self.returns.to(device)
        self.action_log_probs = self.action_log_probs.to(device)
        self.action_cat = self.action_cat.to(device)
        self.action_dia = self.action_dia.to(device)
        self.masks = self.masks.to(device)

    def insert(self, obs, actions, action_log_probs, value_preds, rewards, masks, next_obs):
        self.action_cat[self.step].copy_(actions[0].squeeze())
        self.action_dia[self.step].copy_(actions[1].squeeze())
        self.action_log_probs[self.step].copy_(action_log_probs.squeeze())
        self.value_preds[self.step].copy_(value_preds.squeeze())
        self.rewards[self.step].copy_(rewards.squeeze())
        self.obs[self.step].copy_(obs.squeeze())
        self.next_obs[self.step + 1].copy_(next_obs.squeeze())
        self.masks[self.step + 1].copy_(masks.squeeze())

        self.step = self.step + 1

    # @torchsnooper.snoop()
    @snoop
    def feed_forward_generator(self, advantages, rank):
        mini_batch_size = self.num_steps // self.mini_batch_num
        sampler = BatchSampler(SubsetRandomSampler(range(self.num_steps)), mini_batch_size, drop_last=False)
        for indices in sampler:
            next_indices = [indice + 1 for indice in indices]
            # [index, dim]
            obs_batch = self.obs[indices]
            next_obs_batch = self.obs[next_indices]
            action_cat_batch = self.action_cat[indices]
            action_dia_batch = self.action_dia[indices].transpose(0, 1)
            value_pred_batch = self.value_preds[indices]
            return_batch = self.returns[indices]
            old_action_log_probs_batch = self.action_log_probs[indices]
            advantages_batch = advantages[indices]
            masks_batch = self.masks[indices]
            yield obs_batch, next_obs_batch, (action_cat_batch, action_dia_batch), value_pred_batch, return_batch, \
                  masks_batch, old_action_log_probs_batch, advantages_batch,

**Comment by zasdfgbnm** [comment]
@nuomizai Thanks for the information. I will look into that.

**Comment by zasdfgbnm** [comment]
Hi @nuomizai , I think this is an old bug in PyTorch. See: https://github.com/pytorch/pytorch/pull/15489
Upgrading PyTorch version would help, and I suggest you do so. But if upgrading PyTorch is not an option for you, please let me know.

**Comment by nuomizai** [comment]
Thank you so much~ I will try this as soon as I go to the lab tomorrow!------------------ 原始邮件 ------------------
发件人: "Gao, Xiang"<notifications@github.com>
发送时间: 2019年8月2日(星期五) 晚上10:47
收件人: "zasdfgbnm/TorchSnooper"<TorchSnooper@noreply.github.com>;
抄送: "nuomizai"<769572294@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [zasdfgbnm/TorchSnooper] What dose it mean by REPR FAILED? (#20)



Hi @nuomizai , I think this is an old bug in PyTorch. See: pytorch/pytorch#15489
 Upgrading PyTorch version would help, and I suggest you do so. But if upgrading PyTorch is not an option for you, please let me know.
 
&mdash;

**Comment by zasdfgbnm** [comment]
Hi @nuomizai, I just released a new version (TorchSnooper v0.7) that improves support on these old versions of PyTorch. Could you please do a
```
pip install --upgrade TorchSnooper
```
And see if the problem is solved?

**Comment by nuomizai** [comment]
Thank you so much, @zasdfgbnm ! My problem has been solved by updating torchsnooper to v0.7! Over the past few days, I've tried to update pytorch by installing it from the source, but I met some problem and didn't go on. So the new version can support the old versions (1.0.1) of pytorch perfectly.
Best wishes!
